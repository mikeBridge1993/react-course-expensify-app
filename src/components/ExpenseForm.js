import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';


export default class ExpenseForm extends React.Component {
  constructor (props){
    super(props);
    
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => {
      return {
        description
      }
    });
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => {
      return {
        note
      }
    });
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      console.log(amount)
      this.setState(() => {
        return {
          amount
        }
      });
    }
    
  };
  
  onDateChange = (createdAt) => {
   
    if (createdAt){
      this.setState(() => {
        return {
          createdAt
        }
      });
    }
    
  };

  onFocusedChange = ({ focused }) => {
    this.setState(() => {
      return {
        calendarFocused: focused
      }
    });
  };

  onSubmit = (e) => {
    e.preventDefault();
    console.log(this.state.createdAt.valueOf())
    if (!this.state.description || !this.state.amount){
      this.setState(() => {
      return {
        error: " Please provide description and amount."
      }
      });
    } else {
      console.log("great success");
      this.setState(() => {
        return {
          error: ''
        }
      });

      this.props.onSubmit(
        {
          description: this.state.description,
          amount: parseFloat(this.state.amount, 10)*100,
          createdAt: this.state.createdAt.valueOf(),
          note: this.state.note
        }
      );
    }
  };

  render() {
    return (
      <div className="container-fluid expenses">
      {this.state.error && <div className="alert alert-danger"><strong>Error.</strong> {this.state.error}</div>}
      <form onSubmit={this.onSubmit} className="p-2 rounded border border-secondary">
        <div className="form-group w-90">
          <label>Description</label>
          <input type="text" className="form-control" 
          autoFocus 
          placeholder="Description" 
          value={this.state.description}
          onChange={this.onDescriptionChange}
          />
          <small className="form-text text-muted">Please describe your expense.</small>
        </div>
        <div className="form-group  w-90">
          <label>Amount</label>
          <input type="text" className="form-control" 
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
          />
        </div>
        <div className="form-group w-90">
          <label>Notes:</label>
          <textarea type="text" className="form-control" placeholder="Notes"
          value={this.state.note}
          onChange={this.onNoteChange}
          ></textarea> 
        </div>
        <div className="form-group w-90">
        <label>Date:</label><br />
        <SingleDatePicker
          date={this.state.createdAt} // momentPropTypes.momentObj or null
          onDateChange={this.onDateChange} // PropTypes.func.isRequired
          focused={this.state.calendarFocused} // PropTypes.bool
          onFocusChange={this.onFocusedChange} // PropTypes.func.isRequired
          numberOfMonths={1}
          isOutsideRange={() => { false }}
        />
        </div>
        <button type="submit" className="btn btn-primary">Submit</button>
        
      </form>
      </div>
    )
  };
}

