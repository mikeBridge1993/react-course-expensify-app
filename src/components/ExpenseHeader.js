import React from 'react';

const ExpenseHeader = () => {
  return (
    <div className="container-fluid expenses">
      <hr className="my-0 py-2"/>
      <h1 className="display-4 my-0 py-3">Expenses List</h1>
      <hr className="mt-2 py-2"/>
    </div>
  );
}



export default ExpenseHeader;
