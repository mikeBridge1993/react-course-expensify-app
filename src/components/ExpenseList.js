import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = (props) => {
  return (
    <div className="container-fluid col-10 offset-1">
      <ul className="list-group mb-2 pb-0 d-flex align-items-center">
        <li className="list-group-item rounded expense-list-header">
            <span className="mr-3">Expense</span>
            <span className="mr-3">Amount</span>
            <span className="mr-3">Date</span>
        </li>
      {props.expenses.map((el) => {
          return <ExpenseListItem key={el.id} {...el} createdAt={el.createdAt} />
      })}
      </ul>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}

export default connect(mapStateToProps)(ExpenseList);

