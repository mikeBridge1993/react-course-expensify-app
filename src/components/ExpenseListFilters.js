import React from 'react';
import { connect } from 'react-redux';
import {setTextFilter ,sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate); 
    this.props.setEndDate(endDate); 
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => {
      return {
        calendarFocused
      };
    });
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  onSortChange = (e) => {
    console.log(e.target.value)
    if (e.target.value === "date-high"){
      this.props.sortByDate('high');
    } else if (e.target.value === "date-low") {
      this.props.sortByDate('low');
    } else if (e.target.value === "amount-high") {
      this.props.sortByAmount('high');
    } else if (e.target.value === "amount-low") {
      this.props.sortByAmount('low');
    }
  }

  render () {
    return (
      <div className="container-fluid expenses">
        <div className="input-group w-25">
          <input type="text" className="form-control" value={this.props.filters.text} onChange={this.onTextChange}/>
          <div className="input-group-append">
            <span className="input-group-text"><i className="fa fa-search"></i></span>
          </div>
        </div>
        <div className="w-25 my-2">
          <select className="form-control" 
            onChange={this.onSortChange}>
            <optgroup label="Date">            
              <option value="date-high">Newest first</option>
              <option value="date-low">Oldest first</option>
            </optgroup>
            <optgroup label="Amount">
              <option value="amount-high">Highest first</option>
              <option value="amount-low">Lowest first</option>
            </optgroup>
          </select>
        </div>
        <div className=" w-50 my-2">
            <DateRangePicker
              startDate={this.props.filters.startDate} // 
              startDateId="begginning" // PropTypes.string.isRequired,
              endDate={this.props.filters.endDate} // 
              endDateId="end" // PropTypes.string.isRequired,
              onDatesChange={this.onDatesChange} // 
              focusedInput={this.state.calendarFocused} // 
              onFocusChange={this.onFocusChange} //
              numberOfMonths={1} 
              showClearDates={true} 
              isOutsideRange={() => false} 
            />
          </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};


const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: criteria => dispatch(sortByAmount(criteria)),
    sortByDate: (criteria) => dispatch(sortByDate(criteria))
  };
};


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);