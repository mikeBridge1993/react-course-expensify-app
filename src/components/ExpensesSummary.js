import React from 'react';
import { connect } from 'react-redux';
import getTotalExpenses from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
  return (
    <div className="container-fluid expenses mb-2">
    <span class="badge badge-success mr-5 p-2 w-100">Viewing {props.totalCount} {props.totalCount === 1 ? "expense": "expenses"} totalling {numeral(props.totalAmount/100).format("$0,0.00")}</span> <br/>
    </div>
  );
}

const mapStateToProps = (state, props) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  
  return {
    totalAmount: getTotalExpenses(visibleExpenses),
    totalCount: state.expenses.length
  };
};

export default connect(mapStateToProps)(ExpensesSummary);

