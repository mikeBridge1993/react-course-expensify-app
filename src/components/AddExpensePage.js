import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startAddExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };

  render () {
    return (
      <div className="container-fluid expenses">
        <div className="container-fluid expenses">
          <hr className="my-0 py-2"/>
          <h1 className="display-4 my-0 py-3">Create New Expense</h1>
          <hr className="mt-2 py-2"/>
          <p className="lead">New expense:</p>  
        </div>
        <ExpenseForm onSubmit={this.onSubmit}/>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);