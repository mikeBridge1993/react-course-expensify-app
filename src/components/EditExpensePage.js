import React from 'react';
import { connect } from 'react-redux';
import {editExpense, removeExpense} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.removeExpense({id: this.props.expense.id});
    this.props.history.push("/");
  };

  render () {
    return (
      <div className="container-fluid expenses">
        <div className="container-fluid expenses">
        <hr className="my-0 py-2"/>
        <h1 className="display-4 my-0 py-3">Edit Expense</h1>
        <hr className="mt-2 py-2"/>
        <p className="lead">Expense:</p>  
        </div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
        <div className="container-fluid expenses">
          <button onClick={this.onRemove}
            className="btn ml-0 btn-danger mt-2">
            Delete</button>
          
        </div>
      </div>
      
    );
  };
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
        return expense.id === props.match.params.id;
      }
    )
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);