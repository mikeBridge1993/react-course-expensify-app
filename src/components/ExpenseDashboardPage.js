import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseHeader from './ExpenseHeader';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
  <div className="container-fluid expenses px-0 bg-light">
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;