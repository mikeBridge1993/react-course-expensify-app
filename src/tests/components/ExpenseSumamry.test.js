import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('should correctly render ExpensesSummary with 1 expense', () => {
  const totals = {
    totalAmount: 1000,
    totalCount: 1
  }

  const wrapper = shallow(<ExpensesSummary {...totals}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
  const totals = {
    totalAmount: 1000,
    totalCount: 2
  }
  
  const wrapper = shallow(<ExpensesSummary {...totals}/>);
  expect(wrapper).toMatchSnapshot();
});