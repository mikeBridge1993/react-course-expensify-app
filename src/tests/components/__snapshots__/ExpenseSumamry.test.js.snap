// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div
  className="container-fluid expenses mb-2"
>
  <span
    className="badge badge-success mr-5 p-2 w-100"
  >
    Viewing 
    1
     
    expense
     totalling 
    $10.00
  </span>
   
  <br />
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div
  className="container-fluid expenses mb-2"
>
  <span
    className="badge badge-success mr-5 p-2 w-100"
  >
    Viewing 
    2
     
    expenses
     totalling 
    $10.00
  </span>
   
  <br />
</div>
`;
